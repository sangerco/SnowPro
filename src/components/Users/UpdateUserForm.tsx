import React, { useEffect, useState } from "react";
import { useSelector, useDispatch } from "react-redux";
import { RootState, AppDispatch } from "../../redux/store";
import {
  UserData,
  fetchOneUser,
  updateUser,
} from "../../redux/slices/userSlice";
import {
  Container,
  Form,
  Button,
  Dimmer,
  Loader,
  Header,
} from "semantic-ui-react";

const UpdateUserForm: React.FC = () => {
  const dispatch = useDispatch<AppDispatch>();
  const users = useSelector((state: RootState) => state.users);
  const user = users.user;

  useEffect(() => {
    if (user) {
      dispatch(fetchOneUser(user.username));
    }
  }, [dispatch, user]);

  let initialState: UserData = {
    id: "",
    username: "",
    password: "",
    email: "",
    firstName: "",
    lastName: "",
    isAdmin: false,
    avatar: "",
    bio: "",
    vLinks: [],
    pLinks: [],
    favMountains: [],
  };

  useEffect(() => {
    if (user) {
      setFormData({
        id: user.id,
        username: user.username,
        password: user.password || "",
        email: user.email,
        firstName: user.firstName,
        lastName: user.lastName,
        isAdmin: user.isAdmin,
        avatar: user.avatar || "",
        bio: user.bio || "",
        vLinks: user.vLinks || [],
        pLinks: user.pLinks || [],
        favMountains: user.favMountains || [],
      });
      setPhotos(user.pLinks || []);
      setVideos(user.vLinks || []);
    }
  }, [user]);

  const [formData, setFormData] = useState(initialState);
  const [photos, setPhotos] = useState(user?.pLinks);
  const [videos, setVideos] = useState(user?.vLinks);

  if (users.loading) {
    return (
      <div>
        <Dimmer active>
          <Loader>Loading...</Loader>
        </Dimmer>
      </div>
    );
  }

  if (users.error) {
    return (
      <div>
        <Dimmer>
          <p>Error! User data cannot be found! {`${users.error}`}</p>
        </Dimmer>
      </div>
    );
  }

  const handleSubmit = (e: React.FormEvent<HTMLFormElement>) => {
    e.preventDefault();

    if (formData) {
      dispatch(updateUser(formData));
    }

    setFormData(initialState);
  };

  const handleChange: React.ChangeEventHandler<HTMLInputElement> = (e) => {
    const { name, value } = e.target;

    setFormData({
      ...formData,
      [name]: value,
    });
  };

  const handleTextAreaChange: React.ChangeEventHandler<HTMLTextAreaElement> = (
    e
  ) => {
    const { name, value } = e.target;

    setFormData({
      ...formData,
      [name]: value,
    });
  };

  const handlePhotoChange = (index: number, value: string) => {
    if (formData.pLinks) {
      const updatedPhotoLinks = [...formData.pLinks];
      updatedPhotoLinks[index] = value;
      setPhotos(updatedPhotoLinks);
    }
  };

  const handleAddPhotoInput = () => {
    if (formData.pLinks) {
      setPhotos([...formData.pLinks, ""]);
    }
  };

  const handleRemovePhotoInput = (index: number) => {
    if (formData.pLinks) {
      const updatedPhotoLinks = [...formData.pLinks];
      updatedPhotoLinks.splice(index, 1);
      setPhotos(updatedPhotoLinks);
    }
  };

  const handleVideoChange = (index: number, value: string) => {
    if (formData.vLinks) {
      const updatedVideoLinks = [...formData.vLinks];
      updatedVideoLinks[index] = value;
      setPhotos(updatedVideoLinks);
    }
  };

  const handleAddVideoInput = () => {
    if (formData.vLinks) {
      setVideos([...formData.vLinks, ""]);
    }
  };

  const handleRemoveVideoInput = (index: number) => {
    if (formData.vLinks) {
      const updatedVideoLinks = [...formData.vLinks];
      updatedVideoLinks.splice(index, 1);
      setVideos(updatedVideoLinks);
    }
  };

  if (user) {
    return (
      <div>
        <Container>
          <Header as="h2">{`Update ${user.username}'s Profile`}</Header>
          <Form onSubmit={handleSubmit}>
            <Form.Field>
              <label>Username</label>
              <input
                placeholder={formData.username}
                name="username"
                value={formData.username}
                onChange={handleChange}
              />
            </Form.Field>
            <Form.Field>
              <label>Password</label>
              <input
                type="password"
                placeholder={formData.password}
                name="password"
                value={formData.password}
                onChange={handleChange}
              />
            </Form.Field>
            <Form.Field>
              <label>First Name</label>
              <input
                placeholder={formData.firstName}
                name="firstName"
                value={formData.firstName}
                onChange={handleChange}
              />
            </Form.Field>
            <Form.Field>
              <label>Last Name</label>
              <input
                placeholder={formData.lastName}
                name="lastName"
                value={formData.lastName}
                onChange={handleChange}
              />
            </Form.Field>
            <Form.Field>
              <label>Email</label>
              <input
                type="email"
                placeholder={formData.email}
                name="email"
                value={formData.email}
                onChange={handleChange}
              />
            </Form.Field>
            <Form.Field>
              <label>About You...</label>
              <textarea
                placeholder={formData.bio}
                name="bio"
                value={formData.bio}
                onChange={handleTextAreaChange}
              />
            </Form.Field>
            <Form.Field>
              <label>Avatar Link</label>
              <input
                placeholder={formData.avatar}
                name="avatar"
                value={formData.avatar}
                onChange={handleChange}
              />
            </Form.Field>
            {formData.pLinks &&
              formData.pLinks.map((photo, index) => (
                <Form.Field key={index}>
                  <label>Photos</label>
                  <input
                    value={photo}
                    onChange={(e) => handlePhotoChange(index, e.target.value)}
                  />
                  {formData.pLinks && index === formData.pLinks.length - 1 && (
                    <Button icon="plus" onClick={handleAddPhotoInput} />
                  )}
                  {index > 0 && (
                    <Button
                      icon="minus"
                      onClick={() => handleRemovePhotoInput(index)}
                    />
                  )}
                </Form.Field>
              ))}
            {formData.vLinks &&
              formData.vLinks.map((video, index) => (
                <Form.Field key={index}>
                  <label>Videos</label>
                  <input
                    value={video}
                    onChange={(e) => handleVideoChange(index, e.target.value)}
                  />
                  {formData.vLinks && index === formData.vLinks.length - 1 && (
                    <Button icon="plus" onClick={handleAddVideoInput} />
                  )}
                  {index > 0 && (
                    <Button
                      icon="minus"
                      onClick={() => handleRemoveVideoInput(index)}
                    />
                  )}
                </Form.Field>
              ))}
            <Button type="submit">Submit Changes</Button>
          </Form>
        </Container>
      </div>
    );
  }

  return null;
};

export default UpdateUserForm;
