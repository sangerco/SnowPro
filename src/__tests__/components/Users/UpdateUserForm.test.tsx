/* eslint-disable testing-library/no-node-access */
import React from "react";
import UpdateUserForm from "../../../components/Users/UpdateUserForm";
import { render, fireEvent, screen } from "@testing-library/react";
import { Provider } from "react-redux";
import { BrowserRouter } from "react-router-dom";
import configureStore from "redux-mock-store";
import thunk from "redux-thunk";
import nock from "nock";

const middlewares = [thunk];
const mockStore = configureStore(middlewares);

describe("test Update User Form", () => {
  const initialState = {
    auth: {
      data: {
        id: "11",
      },
    },
    reviews: {
      reviews: [
        {
          id: "review-1",
          userId: "11",
          username: "testuser",
          skiAreaSlug: "ski-area-1",
          skiAreaName: "Ski Area 1",
          header: "test review",
          body: "test review body",
          stars: 5,
          photos: [],
          createdAt: "fake created Date",
          replies: [],
        },
        {
          id: "review-2",
          userId: "22",
          username: "testuser2",
          skiAreaSlug: "ski-area-2",
          skiAreaName: "Ski Area 2",
          header: "test review",
          body: "test review body",
          stars: 2,
          photos: [],
          createdAt: "fake created Date",
          replies: [],
        },
      ],
    },
    reviewReplies: {
      reviewReply: {
        id: "review-reply-1",
        reviewId: "review-1",
        userId: "22",
        username: "testuser2",
        body: "test review reply body",
        slug: "ski-area-1",
        createdAt: "fake created date",
      },
    },
    skiAreas: {
      skiArea: {
        slug: "ski-area-1",
        name: "Ski Area 1",
        country: "US",
        region: "CO",
        href: "www.fakehref.com",
        location: {
          latitude: 0,
          longitude: 0,
        },
        lifts: {},
        twitter: {},
      },
      skiAreas: [
        {
          slug: "ski-area-1",
          name: "Ski Area 1",
          country: "US",
          region: "CO",
          href: "www.fakehref.com",
          location: {
            latitude: 0,
            longitude: 0,
          },
          lifts: {},
          twitter: {},
        },
        {
          slug: "ski-area-1",
          name: "Ski Area 1",
          country: "US",
          region: "CO",
          href: "www.fakehref.com",
          location: {
            latitude: 0,
            longitude: 0,
          },
          lifts: {},
          twitter: {},
        },
      ],
    },
    users: {
      user: {
        id: "11",
        username: "testuser",
        firstName: "Test",
        lastName: "User",
        email: "test@user.com",
        bio: "test user bio",
        avatar: "https://testdata.com/testavatar.jpg",
        isAdmin: false,
        videos: [],
        photos: [],
        favMountains: [],
      },
      users: [
        {
          id: "11",
          username: "testuser",
          firstName: "Test",
          lastName: "User",
          email: "test@user.com",
          bio: "test user bio",
          avatar: "https://testdata.com/testavatar.jpg",
          isAdmin: false,
        },
        {
          id: "22",
          username: "testuser2",
          firstName: "Test",
          lastName: "User2",
          email: "test@user2.com",
          bio: "test user bio",
          avatar: "https://testdata.com/testavatar2.jpg",
          isAdmin: true,
        },
      ],
    },
  };
  const store = mockStore(initialState);
  it("should render without crashing", () => {
    render(
      <Provider store={store}>
        <BrowserRouter>
          <UpdateUserForm />
        </BrowserRouter>
      </Provider>
    );
  });
  it("should match snapshot", () => {
    const { asFragment } = render(
      <Provider store={store}>
        <BrowserRouter>
          <UpdateUserForm />
        </BrowserRouter>
      </Provider>
    );
    expect(asFragment()).toMatchSnapshot();
  });
});

jest.mock("react-router-dom", () => ({
  ...jest.requireActual("react-router-dom"),
  useParams: jest.fn(),
}));

describe("test user form", () => {
  const username = "testuser";
  const useParamsMock = jest.requireMock("react-router-dom")
    .useParams as jest.Mock;
  useParamsMock.mockReturnValue({ username });

  const initialState = {
    auth: {
      data: {
        id: "11",
      },
    },
    reviews: {
      reviews: [
        {
          id: "review-1",
          userId: "11",
          username: "testuser",
          skiAreaSlug: "ski-area-1",
          skiAreaName: "Ski Area 1",
          header: "test review",
          body: "test review body",
          stars: 5,
          photos: [],
          createdAt: "fake created Date",
          replies: [],
        },
        {
          id: "review-2",
          userId: "22",
          username: "testuser2",
          skiAreaSlug: "ski-area-2",
          skiAreaName: "Ski Area 2",
          header: "test review",
          body: "test review body",
          stars: 2,
          photos: [],
          createdAt: "fake created Date",
          replies: [],
        },
      ],
    },
    reviewReplies: {
      reviewReply: {
        id: "review-reply-1",
        reviewId: "review-1",
        userId: "22",
        username: "testuser2",
        body: "test review reply body",
        slug: "ski-area-1",
        createdAt: "fake created date",
      },
    },
    skiAreas: {
      skiArea: {
        slug: "ski-area-1",
        name: "Ski Area 1",
        country: "US",
        region: "CO",
        href: "www.fakehref.com",
        location: {
          latitude: 0,
          longitude: 0,
        },
        lifts: {},
        twitter: {},
      },
      skiAreas: [
        {
          slug: "ski-area-1",
          name: "Ski Area 1",
          country: "US",
          region: "CO",
          href: "www.fakehref.com",
          location: {
            latitude: 0,
            longitude: 0,
          },
          lifts: {},
          twitter: {},
        },
        {
          slug: "ski-area-1",
          name: "Ski Area 1",
          country: "US",
          region: "CO",
          href: "www.fakehref.com",
          location: {
            latitude: 0,
            longitude: 0,
          },
          lifts: {},
          twitter: {},
        },
      ],
    },
    users: {
      user: {
        id: "11",
        username: "testuser",
        firstName: "Test",
        lastName: "User",
        email: "test@user.com",
        bio: "test user bio",
        avatar: "https://testdata.com/testavatar.jpg",
        isAdmin: false,
        videos: [],
        photos: [],
        favMountains: [],
      },
      users: [
        {
          id: "11",
          username: "testuser",
          firstName: "Test",
          lastName: "User",
          email: "test@user.com",
          bio: "test user bio",
          avatar: "https://testdata.com/testavatar.jpg",
          isAdmin: false,
        },
        {
          id: "22",
          username: "testuser2",
          firstName: "Test",
          lastName: "User2",
          email: "test@user2.com",
          bio: "test user bio",
          avatar: "https://testdata.com/testavatar2.jpg",
          isAdmin: true,
        },
      ],
    },
  };
  const store = mockStore(initialState);

  it("should render a form", async () => {
    nock("http://localhost:5000")
      .defaultReplyHeaders({
        "access-control-allow-origin": "*",
      })
      .get(`/users/${username}`)
      .reply(200, {
        id: "11",
        username: "testuser",
        firstName: "Test",
        lastName: "User",
        email: "test@user.com",
        password: "password",
        avatar: "https://testlinks.com/testavatar.jpg",
        bio: "test user bio",
        isAdmin: "false",
        videos: [],
        photos: [],
        favMountains: [],
      });
    nock("http://localhost:5000")
      .defaultReplyHeaders({
        "access-control-allow-origin": "*",
      })
      .get("/ski-areas")
      .reply(200, []);
    nock("http://localhost:5000")
      .defaultReplyHeaders({
        "access-control-allow-origin": "*",
      })
      .patch("/api/users/testuser")
      .reply(200, {
        id: "11",
        username: "testuser",
        firstName: "Test",
        lastName: "User",
        email: "test@user.com",
        password: "password",
        avatar: "https://testlinks.com/testavatar.jpg",
        bio: "updated test user bio",
        isAdmin: "false",
        videos: [],
        photos: [],
        favMountains: [],
      });

    render(
      <Provider store={store}>
        <BrowserRouter>
          <UpdateUserForm />
        </BrowserRouter>
      </Provider>
    );

    const form = screen.getByTestId("update-user-form");
    expect(form).toBeInTheDocument();

    const usernameField = screen.getByPlaceholderText("testuser");
    expect(usernameField).toBeInTheDocument();

    const firstNameField = screen.getByPlaceholderText("Test");
    expect(firstNameField).toBeInTheDocument();

    const button = screen.getByTestId("submit-button");
    expect(button).toBeInTheDocument();
  });
});
