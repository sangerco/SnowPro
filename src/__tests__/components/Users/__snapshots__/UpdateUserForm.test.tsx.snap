// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test Update User Form should match snapshot 1`] = `
<DocumentFragment>
  <div
    class="ui active transition visible dimmer"
    style="display: flex;"
  >
    <div
      class="content"
    >
      <h1
        class="ui orange header"
      >
        Access not authorized.
      </h1>
      <button
        class="ui red button"
      >
        Go Back
      </button>
    </div>
  </div>
</DocumentFragment>
`;
